File: LinkedListSerialize/ListRand.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;

namespace LinkedListSerialize
{
  class ListRand
  {
    public ListNode Head;
    public ListNode Tail;
    public int Count = 0;

    public void Serialize(StreamWriter writer)
    {
      Dictionary<ListNode, int> dictionaryNodes = new Dictionary<ListNode, int>();
      var counter = 0;
      var pointer = Head;
      while (pointer != null)
      {
        dictionaryNodes[pointer] = counter;
        ++counter;
        pointer = pointer.Next;
      }
      pointer = Head;
      while (pointer != null)
      {
        writer.WriteLine(pointer.Data);
        if (pointer.Rand != null)
        {
          writer.WriteLine(dictionaryNodes[pointer.Rand]);
        }
        else
        {
          writer.WriteLine(-1);
        }
        pointer = pointer.Next;
      }
    }

    public void Deserialize(StreamReader reader)
    {
      Dictionary<ListNode, int> dictionaryNodes = new Dictionary<ListNode, int>();
      string line;
      while ((line = reader.ReadLine()) != null)
      {
        var node = Add(line);
        line = reader.ReadLine();
        dictionaryNodes[node] = Int32.Parse(line);
      }
      var pointer = Head;
      while (pointer != null)
      {
        if (dictionaryNodes[pointer] != (-1))
        {
          pointer.Rand = dictionaryNodes.ElementAt(dictionaryNodes[pointer]).Key;
        }
        pointer = pointer.Next;
      }
    }

    public ListNode Add(string value)
    {
      ListNode node = new ListNode();
      node.Data = value;
      if (Tail != null)
      {
        Tail.Next = node;
      }

      node.Prev = Tail;
      node.Rand = null;

      Tail = node;
      if (Head == null)
      {
        Head = node;
      }
      ++Count;
      return node;
    }

    public void AddRandomLink(int indexNode, int indexTarget)
    {
      if (((indexNode < 0) || (indexNode > Count)) ||
          ((indexTarget < 0) || (indexTarget > Count)))
      {
        System.Console.WriteLine("Index out of bounds list");
        return;
      }

      ListNode pointerNode = null;
      ListNode pointerTarget = null;
      var pointer = Head;
      int counter = 0;
      while (pointer != null)
      {
        if (indexNode == counter)
        {
          pointerNode = pointer;
        }
        if (indexTarget == counter)
        {
          pointerTarget = pointer;
        }
        if ((pointerNode != null) && (pointerTarget != null))
        {
          break;
        }
        pointer = pointer.Next;
        ++counter;
      }
      pointerNode.Rand = pointerTarget;
    }

    public void Print()
    {
      var node = Head;
      while (node != null)
      {
        System.Console.Write(node.Data + " ");
        if (node.Rand != null)
        {
          System.Console.Write("-(" + node.Rand.Data + ") ");
        }
        node = node.Next;
      }
      System.Console.WriteLine();
    }
  }
}

File: SerializedLinkedList/ListRand.cpp
#include "ListRand.h"
#include <iostream>
#include <unordered_map>

ListRand::ListRand() {
	Head = nullptr;
	Tail = nullptr;
	Count = 0;
}

ListRand::~ListRand() {
	auto pointer = Head;
	while (pointer) {
		auto node = pointer;
		pointer = pointer->Next;
		delete node;
		node = nullptr;
	}
}

ListNode* ListRand::Add(const std::string& value) {
	ListNode* node = new ListNode();
	node->Data = value;
	if (Tail) {
		Tail->Next = node;
	}
	node->Prev = Tail;
	node->Rand = nullptr;

	Tail = node;
	if (!Head) {
		Head = node;
	}
	++Count;
	return node;
}

void ListRand::AddRandomLink(int nodeIndex, int targetIndex) {
	if (((nodeIndex < 0) || (nodeIndex > Count)) ||
		  ((targetIndex < 0) || (targetIndex > Count))) {
		std::cout << "Index out of bounds list" << std::endl;
		return;
	}
	ListNode* nodePointer = nullptr;
	ListNode* targetPointer = nullptr;
	int counter = 0;
	auto pointer = Head;
	while (pointer) {
		if (nodeIndex == counter) {
			nodePointer = pointer;
		}
		if (targetIndex == counter) {
			targetPointer = pointer;
		}
		if ((nodePointer) && (targetPointer)) {
			nodePointer->Rand = targetPointer;
			break;
		}
		pointer = pointer->Next;
		++counter;
	}
}

void ListRand::Print() {
	auto pointer = Head;
	while (pointer) {
		std::cout << pointer->Data << " ";
		if (pointer->Rand) {
			std::cout << "-(" << pointer->Rand->Data << ") ";
		}
		pointer = pointer->Next;
	}
	std::cout << std::endl;
}

void ListRand::Serialize(std::ofstream &out) {
	std::unordered_map<ListNode*, int> nodeMap;
	int counter = 0;
	auto pointer = Head;
	while (pointer) {
		nodeMap.insert({ pointer, counter });
		++counter;
		pointer = pointer->Next;
	}
	pointer = Head;
	while (pointer) {
		out << pointer->Data << std::endl;
		if (pointer->Rand)
		{
			out << nodeMap[pointer->Rand] << std::endl;
		}
		else
		{
			out << -1 << std::endl;
		}
		pointer = pointer->Next;
	}
}

void ListRand::Deserialize(std::ifstream &in) {
	std::unordered_map<ListNode*, int> nodeMap;
	std::unordered_map<int, ListNode*> nodeIndexMap;
	std::string line;
	int counter = 0;
	while (getline(in, line)) {
		auto node = Add(line);
		getline(in, line);
		nodeMap.insert({ node, std::stoi(line) });
		nodeIndexMap.insert({ counter, node });
		++counter;
	}
	auto pointer = Head;
	while (pointer) {
		if (nodeMap[pointer] != (-1)) {
			pointer->Rand = nodeIndexMap[nodeMap[pointer]];
		}
		pointer = pointer->Next;
	}
}